商品出品機能

sisyphe0713 opened the pull request:
# What: 商品出品機能
# Why: ユーザーが商品を出品できるようにするため

ログイン状態のユーザーだけが、商品出品ページへ遷移できる
https://gyazo.com/47761eda4e94a8ec4e7917d727047cc2

ログアウト状態のユーザーは、商品出品ページへ遷移しようとすると、ログインページへ遷移する
https://gyazo.com/536cfbccccc14e718063fd8feb255910 (修正前)
https://gyazo.com/691f997e447768faf5d1588c895f6f55 (修正後)
https://gyazo.com/3c96434ccd2a3538f79e542038ba8c8b (修正後2)

入力された販売価格によって、販売手数料や販売利益の表示が変わる
https://gyazo.com/b5a440d86f08198562e2d040c1be9d65

必要な情報を適切に入力すると、商品の出品ができる
https://gyazo.com/b1be3b20dd2703e6814ea6fe5453b43e

入力に問題がある状態で出品ボタンが押されたら、出品ページに戻りエラーメッセージが表示される
https://gyazo.com/c06ba82a4ab6972440bbe14c9d79bc9c

テストコード実行結果
https://gyazo.com/b85da9474e74cc46215d799f3bceb80d (修正前)
https://gyazo.com/f10664674d7432dd39c77e823f37506a (修正後)

sisyphe0713 committed:
商品投稿ページへの遷移

-newアクションのルーティング設定
-newアクションをコントローラーに定義
-投稿画面へのパスをlink_toに設定

sisyphe0713 committed:
商品データの保存

-createアクションのルーティング設定
-createアクションをコントローラーに定義
-バリデーションを設定

sisyphe0713 committed:
Active Storageと画像加工ツールの導入

-Active Storageをインストール
-Mini Magickをインストール
-Image Processingをインストール

sisyphe0713 committed:
画像データ保存可能に

-アソシエーション定義
-ストロングパラメーター編集

sisyphe0713 committed:
商品出品データをテーブルに保存

-ActiveHash導入
-category, status, delivery_fee, prefecture, ship_dayをActiveHashを用いて記述

sisyphe0713 committed:
ログインの有無による遷移分岐

sisyphe0713 committed:
販売手数料、販売利益の自動計算

sisyphe0713 committed:
テストコード

-バリデーション(販売価格が数字で300〜9999999の範囲である)追加
-テストコード記述

aoitech commented on app/controllers/items_controller.rb:
```diff
@@ -1,4 +1,31 @@
   end
+
+  def new
+    if user_signed_in? != true
```
```ruby
if user_signed_in? != true
 redirect_to root_path
```
こちらの記述ですと、ユーザーがログインしていない場合にトップページにリダイレクトしているかと存じます。
trelloの要件に`ログアウト状態のユーザーは、商品出品ページへ遷移しようとすると、ログインページへ遷移すること`とございますので、記述を変更しましょう。

deviseの`authenticate_user!`というメソッドを使用するとコードの記述量が少なく実装できるかと存じます。

ご不明点がございましたら、コネクトでご質問いただけますと幸いです。
お手数をお掛けいたしますが、よろしくお願いいたします。

sisyphe0713 replied:
ご指摘ありがとうございます。
遷移先うっかりしていました。ログインページに遷移するように修正しました。
https://gyazo.com/691f997e447768faf5d1588c895f6f55

authenticate_user!は使おうとしたのですが、ホームが表示されなくて時新規登録ができなくなってしまうんですよね…どういう実装を想定しているのか分からないです…。

Katsuno0814 replied:
https://github.com/sisyphe0713/furima-30355/pull/3#discussion_r565753144

before_action :authenticate_user!と記述してしまうと全てのメソッドに対して設定されてしまい
ホームが表示されなくなるかと思います。以下の様にすることでそれを防ぐことができます！

```.rb
before_action :authenticate_user!, only: [:new, :create]
```

aoitech commented on app/controllers/items_controller.rb:
```diff
@@ -1,4 +1,31 @@
+  end
+
+  def create
+    # Item.create(item_params)
```
```ruby
# Item.create(item_params)
```
こちらのコメントアウトですが、不要であれば削除しましょう。
もし必要であれば、第三者でもわかるように「残す目的」「なぜ必要か」という理由などを、コメントとして書いておくことをおすすめします。

sisyphe0713 replied:
ご指摘ありがとうございます。
削除しました。

aoitech commented on app/views/items/new.html.erb:
```diff
@@ -5,10 +5,10 @@
-    <%= form_with local: true do |f| %>
+    <%= form_with model: @item, local: true do |f| %>
 
     <%# インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
```
```ruby
<%# インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
```
こちらのコメントアウトも、不要であれば削除しておきましょう。
もし必要であれば、第三者でもわかるように「残す目的」「なぜ必要か」という理由などを、コメントとして書いておくことをおすすめします。
他にも不要なコメントアウトがございましたら、削除しておきましょう。

sisyphe0713 replied:
ご指摘ありがとうございます。
削除しました。

aoitech commented on spec/factories/categories.rb:
```diff
@@ -0,0 +1,5 @@
+FactoryBot.define do
```
こちらのファイルは使用されていますでしょうか？
使用されないのであれば、削除しましょう。
他にも使用されないテストファイルがございましたら、削除しておきましょう。

sisyphe0713 replied:
ご指摘ありがとうございます。
ActiveHashで扱っているモデルのテストファイルは全て削除しました。

aoitech commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,89 @@
+        expect(@item.errors.full_messages).to include("Delivery fee can't be blank")
+      end
+
+      it "配送元の地域の情報がないと登録できない" do
```
下記のテストコードの追記をお願いいたします。

商品の状態, 配送料の負担, 発送元の地域,  発送までの日数, カテゴリー
```
・1( '--')以外でないと登録できないこと
```

価格
```
・半角英数混合では登録できないこと
・半角英語では登録できないこと
```

sisyphe0713 replied:
ご指摘ありがとうございます。
テストコード追記しました。

実行結果です↓(書いていただいたものと順序が異なりますが、最後7つが該当のテストです)
https://gyazo.com/f10664674d7432dd39c77e823f37506a

aoitech reviewed (commented):
レビュー依頼ありがとうございます。
5点コメントいたしましたので、ご確認いただけますと幸いです。
よろしくお願いいたします。

sisyphe0713 committed:
レビューコメント対応

-ログインしていない時の遷移先がログインページでなくホームになっていたバグを解消
-不要なコメントアウトを削除
-不要なテストファイルを削除
-テストコード追加

Katsuno0814 reviewed (commented):
レビュー依頼ありがとうございます！
いただきましたコメントにお返事しましたので、ご確認お願いいたします！

sisyphe0713 commented:
ご指摘ありがとうございます。
なるほど、only:を使えばよかったのですね。
ありがとうございます！

修正しました。表示されるエラーの表現も異なりますね。
https://gyazo.com/3c96434ccd2a3538f79e542038ba8c8b

sisyphe0713 committed:
レビューコメント対応2

-authenticate_user!を用いて簡潔化

NaturalRadio commented on app/models/item.rb:
```diff
@@ -1,2 +1,28 @@
+
+  with_options presence: true do
+    validates :name
+    validates :price, :numericality => {:only_integer => true, :greater_than_or_equal_to	=> 300, :less_than_or_equal_to => 9999999}
```
```
    validates :price, :numericality => {:only_integer => true, :greater_than_or_equal_to	=> 300, :less_than_or_equal_to => 9999999}
```
こちら`ハッシュロケット`を使用されていますので修正してみましょう！

sisyphe0713 replied:
ご指摘ありがとうございます。
修正しました。

NaturalRadio commented on spec/factories/items.rb:
```diff
@@ -0,0 +1,17 @@
+  factory :item do
+    association :user
+    name {"item"}
+    price	{"300"}
+    text {Faker::Lorem.sentence}
+    category_id	{"2"}
+    status_id	{"2"}
+    delivery_fee_id	{"2"}
+    prefecture_id	{"2"}
+    ship_day_id {"2"}
```
こちらは数値として定義をしておきましょう！
※`"(ダブルクォーテーション)`に着目すると実装がスムーズかと思われます！

sisyphe0713 replied:
ご指摘ありがとうございます。
修正しました。

テストコードでは私自身意識せず数値で扱っていましたが、
半角数字で入力したものは文字列ではなく数値として扱われているのですね。

NaturalRadio reviewed (commented):
レビュー依頼ありがとうございます！
コメントしましたので、ご確認をお願いいたします！

sisyphe0713 committed:
レビューコメント対応3

-ハッシュロケットをシンボルに
-半角数字の文字列を数値に

chisesakai commented on app/views/items/new.html.erb:
```diff
@@ -99,7 +97,7 @@
+          <%= f.text_field :price, class:"price-input", id:"item-price", placeholder:"例）300" %>
         </div>
         <div class="price-content">
           <span>販売手数料 (10%)</span>
```
**こちらの実装は任意です**
現状、JSでの価格表示部分で「¥」マークと「円」が両方表示されておりますので、htmlの「円」の部分を削除すると良いかと思います。

sisyphe0713 replied:
ご指摘ありがとうございます。
おっしゃる通りですね、修正しました。

chisesakai reviewed (commented):
#### LGTMです
レビュー依頼ありがとうございます！
LGTMとさせていただきます！実装お疲れさまでした！
下記のみ、プルリクエスト内で修正お願い致します。
#### マークダウン記法を反映させましょう
こちらのプルリクエストですが、マークダウン記法が反映されておりません。
#とwhatの間に半角スペース（例:`# what`）を入れると、

# what
上記のようにマークダウン記法が反映されます。
今のうちにマスターするため、こちらのプルリクエストに用いてみましょう。

こちらのブランチは忘れずにMergeいただき、ローカルのMasterにPullした後に次の実装のためのブランチを切るようにしましょう。
引き続き実装頑張ってください！




また、1件コメントいたしましたが、指摘に関しては任意実装のため、修正される場合も新たにレビュー依頼をしていただく必要はありません。
参考程度にご確認ください！

sisyphe0713 committed:
Minor revision

-円と¥の重複を削除

sisyphe0713 commented:
LGTMありがとうございます。

マークダウン記法のこともありがとうございます。
面白いですね。

皆様のコメントのおかげで、自分では気づくことのなかったことを気づいたり理解することができて、とても楽しかったです。
どうもありがとうございました！
